chain/chunks/src/lib.rs

```diff
@@ -1075,6 +1092,143 @@ impl ShardsManager {
         )
     }
 
+    /// Prepares response to a partial encoded chunk request from a chunk read
+    /// from the storage.
+    ///
+    /// This requires encoding the chunk and as such is computationally
+    /// expensive operation.  If possible, the request should be served from
+    /// EncodedChunksCacheEntry or PartialEncodedChunk instead.
+    #[cfg_attr(feature = "test_features", visibility::make(pub))]
+    fn prepare_partial_encoded_chunk_response_from_chunk(
+        &mut self,
+        request: PartialEncodedChunkRequestMsg,
+        chain_store: &mut ChainStore,
+        rs: &mut ReedSolomonWrapper,
+        chunk: ShardChunk,
+    ) -> Option<PartialEncodedChunkResponseMsg> {
+        // TODO(mina86): Is it correct to use rs.data_shard_count() and
+        // rs.total_shard_count() here?  Can those values change over time?
+        let total_parts = rs.total_shard_count();
+        for ord in request.part_ords.iter() {
+            let ord: usize = (*ord).try_into().unwrap();
+            if ord >= total_parts {
+                debug!(target:"chunks", "Not sending {:?}, requested part_ord {} but we only expect {} total",
+                       request.chunk_hash.0, ord, total_parts);
+                return None;
+            }
+        }
+
+        // Figure out basic parameters such as shard id, epoch id, protocol
+        // version and shard layout.  Those will be needed later when fetching
+        // other data and constructing encoded shard chunk.
+        let header = chunk.cloned_header();
+        let prev_block_hash = header.prev_block_hash_ref();
+        let shard_id = header.shard_id();
+        let shard_layout = self
+            .runtime_adapter
+            .get_shard_layout_from_prev_block(prev_block_hash)
+            .map_err(|err| {
+                debug!(target: "chunks", "Not sending {:?}, failed to get shard layout: {}", request.chunk_hash.0, err);
+            })
+            .ok()?;
+
+        // Get outgoing receipts for the chunk and construct vector of their
+        // proofs.
+        let outgoing_receipts = chain_store
```

Comment:
Here you can just use `chunk.receipts()` instead of reading from store, since we already have the chunk content